screenkey: init at 0.9

rasendubi opened the pull request:
###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

rasendubi committed:
screenkey: init at 0.9

dotlambda commented on pkgs/top-level/all-packages.nix:
```diff
@@ -4514,6 +4514,10 @@ with pkgs;
     quazip = quazip_qt4;
   };
 
+  screenkey = callPackage ../applications/video/screenkey {
```
Better use `python2Packages.callPackage`. Then you can leave out line 4518.

dotlambda commented on pkgs/applications/video/screenkey/default.nix:
```diff
@@ -0,0 +1,42 @@
+  meta = with lib; {
+    homepage = https://www.thregr.org/~wavexx/software/screenkey/;
+    description = "A screencast tool to display your keys inspired by Screenflick";
+    license = licenses.gpl3;
```
This should be `gpl3Plus`.

rasendubi commented:
@dotlambda, thanks for the review—fixed!

dotlambda commented on pkgs/applications/video/screenkey/default.nix:
```diff
@@ -0,0 +1,42 @@
+  buildInputs = [
+    distutils_extra
+    setuptools-git
+    intltool
```
Why didn't you include all dependencies from https://www.thregr.org/~wavexx/software/screenkey/#installation-and-basic-usage?
Also, some of these should go into `nativeBuildInputs` since they are only needed for building. See https://nixos.org/nixpkgs/manual/#ssec-stdenv-attributes.

rasendubi replied:
`slop` is an optional runtime dependency and is not required for normal operation. (Even "install dependencies" command for Ubuntu/Debian does not list it.)

As for `nativeBuildInputs`, I agree. (Developed the expression on top of the stable channel and it does not feature `nativeBuildInputs` as a separate argument in `mk-python-derivation` yet.)

dotlambda replied:
:+1:

dotlambda commented on pkgs/applications/video/screenkey/default.nix:
```diff
@@ -0,0 +1,42 @@
+buildPythonApplication rec {
+  name = "screenkey-0.9";
+
+  src = fetchFromGitHub {
```
Why don't you use https://www.thregr.org/~wavexx/software/screenkey/releases/screenkey-0.9.tar.gz?
Also, you could specify `pname` and `version` and leave out `name`.
Sorry for the many comments :)

rasendubi replied:
IDK, `fetchFromGitHub` is more convenient to use. GitHub is also less likely to die or move.

Don't see a point in using `pname` and `version`—I am using `name` as `rev` (and reconstructing it from `pname` and `version` would remove all the benefit)

dotlambda replied:
Alright, as you wish.

FRidh replied:
Please do use `pname` and `version` since in long term `buildPython*` will likely get rid of `name`.

rasendubi replied:
ok, fixed

FRidh commented on pkgs/applications/video/screenkey/default.nix:
```diff
@@ -0,0 +1,42 @@
+  ];
+
+  makeWrapperArgs = [
+    "--set LD_LIBRARY_PATH ${libX11}/lib:${libXtst}/lib"
```
Have you considered patching instead?

rasendubi replied:
Do you mean `patchelf`? If so, there is no ELF to patch.

https://github.com/wavexx/screenkey/blob/1b9d6c2dc1e4e9c39d85d756796bc8f54febeb97/Screenkey/xlib.py#L9

jtojnar replied:
See for example here https://github.com/NixOS/nixpkgs/blob/92513b8a41fc4d1d7863c803be9aa1f3746cef42/pkgs/tools/admin/salt/fix-libcrypto-loading.patch#L10

jtojnar replied:
You would create `paths.patch`

```patch
--- a/Screenkey/xlib.py
+++ b/Screenkey/xlib.py
@@ -6,7 +6,7 @@
 from ctypes import *
 
 ## base X11
-libX11 = CDLL('libX11.so.6')
+libX11 = CDLL('@libX11@/lib/libX11.so.6')
 
 # types
 Atom = c_ulong
@@ -278,7 +278,7 @@
 
 
 ## record extensions
-libXtst = CDLL('libXtst.so.6')
+libXtst = CDLL('@libXtst@/lib/libXtst.so.6')
 
 # types
 XPointer = String

```

Then you would add 

```nix
{
patches = [
  (substituteAll {
    src = ./paths.patch;
    inherit libX11 libXtst;
  })
];
}
```

FRidh replied:
`.so` can be replaced with `${stdenv.hostPlatform.extensions.sharedLibrary}`

grahamc commented:
@GrahamcOfBorg build screenkey

GrahamcOfBorg reviewed (commented):
Failure for system: x86_64-darwin

```
Package ‘screenkey-0.9’ in /Users/graham/nix-borg/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/grahamc-zoidberg/pkgs/applications/video/screenkey/default.nix:38 is not supported on ‘x86_64-darwin’, refusing to evaluate.

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowBroken = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowBroken = true; }
to ~/.config/nixpkgs/config.nix.

```

GrahamcOfBorg reviewed (commented):
Success for system: x86_64-linux

```
reading manifest template 'MANIFEST.in'
warning: no previously-included files found matching '.gitignore'
writing manifest file 'screenkey.egg-info/SOURCES.txt'
running build_ext

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
/nix/store/n55akhflgwfcvr9636hzv5akqh4kir09-screenkey-0.9
```

FRidh commented:
```
Ran 0 tests in 0.000s
```
No tests are found. Either the test runner cannot find tests and needs to be patched, or there simply are no tests in which case the tests need to be disabled. Do include a comment explaining why the tests are disabled.

rasendubi commented:
There are no tests.

Why should the check phase be disabled? It is not failing and runs within a second so it does no harm. Furthermore, if upstream adds tests, it's very likely the check phase won't be re-enabled. Am I missing something?

GrahamcOfBorg reviewed (commented):
Success for system: aarch64-linux

```
reading manifest template 'MANIFEST.in'
warning: no previously-included files found matching '.gitignore'
writing manifest file 'screenkey.egg-info/SOURCES.txt'
running build_ext

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
/nix/store/rvqy26hbac7xy798z1kiavjy85hnzgj8-screenkey-0.9
```

rasendubi commented:
Replaced the wrapper hack with a patch to screenkey.

@GrahamcOfBorg build screenkey

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34167&attempt_id=43329107-a0a8-40a9-a9c9-3984d31162d4)

<details><summary>Partial log (click to expand)</summary><p>

```
reading manifest template 'MANIFEST.in'
warning: no previously-included files found matching '.gitignore'
writing manifest file 'screenkey.egg-info/SOURCES.txt'
running build_ext

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
/nix/store/by7l0zznla90z0wfapp0gyndn0ylqdk4-screenkey-0.9
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.gsc.io/?key=nixos/nixpkgs.34167&attempt_id=6bd9700e-de33-4a38-b1f4-eb309be055a1)

<details><summary>Partial log (click to expand)</summary><p>

```
reading manifest template 'MANIFEST.in'
warning: no previously-included files found matching '.gitignore'
writing manifest file 'screenkey.egg-info/SOURCES.txt'
running build_ext

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
/nix/store/sgjv2d5xlang64xpq34ncakvgmhzi39r-screenkey-0.9
```
</p></details>

dotlambda commented:
The comment should be added in order not to confuse other contributors. When I see "Ran 0 tests in 0.000s", I sometimes try to look up what the correct `checkPhase` would be, in this case just to find that there are no tests. If you disable tests and add a comment, everyone will know that it's fine like that.

jtojnar commented:
Apparently, we already have screenkey in `pkgs/top-level/python-packages.nix ` could you change the the derivation there to `pkgs.screenkey` as a legacy alias?

Mic92 commented on pkgs/applications/video/screenkey/default.nix:
```diff
@@ -0,0 +1,50 @@
+  nativeBuildInputs = [
+    distutils_extra
+    setuptools-git
+    intltool
```
Adding `wrapGAppsHook` in `nativeBuildInputs` and `defaultIconTheme`, `hicolor_icon_theme` in `buildInputs` should provide icons regardless of the desktop environments. The other pull request for screenkey seems to do something similar: https://github.com/NixOS/nixpkgs/pull/34898

rasendubi committed:
pythonPackages.screenkey: remove

rasendubi commented:
@Mic92, updated to include icons.

@jtojnar, I removed `pythonPackages.screenkey`. (Don't think it deserves an alias as I believe nobody is using it.)

@vidbina would you like to review/test as well?

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34167&attempt_id=6b707828-2744-40c3-ad79-cbb16d8f0137)

<details><summary>Partial log (click to expand)</summary><p>

```
hicolorPreFixupPhase
post-installation fixup
Wrapping program /nix/store/f7884fmzbbbrmlqvjcrkxgb0lnkpm7yn-screenkey-0.9/bin/screenkey
shrinking RPATHs of ELF executables and libraries in /nix/store/f7884fmzbbbrmlqvjcrkxgb0lnkpm7yn-screenkey-0.9
strip is /nix/store/5qj61lcvzlap87rf6blvf8p577d482bv-binutils-2.28.1/bin/strip
stripping (with command strip and flags -S) in /nix/store/f7884fmzbbbrmlqvjcrkxgb0lnkpm7yn-screenkey-0.9/lib  /nix/store/f7884fmzbbbrmlqvjcrkxgb0lnkpm7yn-screenkey-0.9/bin 
patching script interpreter paths in /nix/store/f7884fmzbbbrmlqvjcrkxgb0lnkpm7yn-screenkey-0.9
checking for references to /tmp/nix-build-screenkey-0.9.drv-0 in /nix/store/f7884fmzbbbrmlqvjcrkxgb0lnkpm7yn-screenkey-0.9...
wrapping `/nix/store/f7884fmzbbbrmlqvjcrkxgb0lnkpm7yn-screenkey-0.9/bin/.screenkey-wrapped'...
/nix/store/f7884fmzbbbrmlqvjcrkxgb0lnkpm7yn-screenkey-0.9
```
</p></details>

jtojnar commented on pkgs/applications/video/screenkey/default.nix:
```diff
@@ -0,0 +1,58 @@
+  propagatedBuildInputs = [
+    pygtk
+
+    defaultIconTheme
```
The icon themes do not need to be propagated, actually the `hicolor_icon_theme` is only used for setup hook.

rasendubi replied:
Oops. Wrong section.

rasendubi replied:
Fixed.

jtojnar commented on pkgs/applications/video/screenkey/default.nix:
```diff
@@ -0,0 +1,60 @@
+
+  buildInputs = [
+    defaultIconTheme
+    hicolor_icon_theme
```
Maybe also add comment that it is added for setup hook – or remove it, since it is propagated by `adwaita-icon-theme`.

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34167&attempt_id=8385b944-ddef-4f7f-854e-ffff759f75bd)

<details><summary>Partial log (click to expand)</summary><p>

```
glibPreFixupPhase
post-installation fixup
Wrapping program /nix/store/bqq6bdjf4d60bzs6f6ap1wskjphwh2k0-screenkey-0.9/bin/screenkey
shrinking RPATHs of ELF executables and libraries in /nix/store/bqq6bdjf4d60bzs6f6ap1wskjphwh2k0-screenkey-0.9
strip is /nix/store/5qj61lcvzlap87rf6blvf8p577d482bv-binutils-2.28.1/bin/strip
stripping (with command strip and flags -S) in /nix/store/bqq6bdjf4d60bzs6f6ap1wskjphwh2k0-screenkey-0.9/lib  /nix/store/bqq6bdjf4d60bzs6f6ap1wskjphwh2k0-screenkey-0.9/bin 
patching script interpreter paths in /nix/store/bqq6bdjf4d60bzs6f6ap1wskjphwh2k0-screenkey-0.9
checking for references to /tmp/nix-build-screenkey-0.9.drv-0 in /nix/store/bqq6bdjf4d60bzs6f6ap1wskjphwh2k0-screenkey-0.9...
wrapping `/nix/store/bqq6bdjf4d60bzs6f6ap1wskjphwh2k0-screenkey-0.9/bin/.screenkey-wrapped'...
/nix/store/bqq6bdjf4d60bzs6f6ap1wskjphwh2k0-screenkey-0.9
```
</p></details>

GrahamcOfBorg commented:
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34167&attempt_id=491a9e05-c989-4d32-b170-5b5acea6d50c)

<details><summary>Partial log (click to expand)</summary><p>

```
   Compiling rustc_mir v0.0.0 (file:///build/rustc-1.22.1-src/src/librustc_mir)
   Compiling rustc_passes v0.0.0 (file:///build/rustc-1.22.1-src/src/librustc_passes)
building of '/nix/store/1q00yw9nphvwvxyk6imxny1rk76rwqy1-rustc-1.22.1.drv' timed out after 3200 seconds
cannot build derivation '/nix/store/i63b6jfgcjz1qkbwsljw4y9vrdsq0qr3-cargo-0.23.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/54gwpxhrss33zvavc8qpv9kvca66ja7j-librsvg-2.42.2.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/4sgr9nibc20a50wqj0m0slhxnb5wsy2d-hook.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/ww752cmphyqnpsqafrhdmq5yzqh3l8v2-icon-naming-utils-0.8.90.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/b1lgn81fw7nxik79zjwshvggx28qgi14-adwaita-icon-theme-3.26.1.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/hvnmgk51fah6qy6fbzj4mi1winpna3yw-screenkey-0.9.drv': 2 dependencies couldn't be built
error: build of '/nix/store/hvnmgk51fah6qy6fbzj4mi1winpna3yw-screenkey-0.9.drv' failed
```
</p></details>

rasendubi commented:
Ok, guys. I went through, like, 6-8 review rounds, and I have little energy to move it forward this way.

If there are no blockers, I would like to get this merged as is. If you have any improvements in mind, feel free to push into my branch.

GrahamcOfBorg commented:
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.34167&attempt_id=5e754f63-7492-4b7f-ace9-21fef5e9a055)

<details><summary>Partial log (click to expand)</summary><p>

```
   Compiling rustc_borrowck v0.0.0 (file:///build/rustc-1.22.1-src/src/librustc_borrowck)
   Compiling rustc_plugin v0.0.0 (file:///build/rustc-1.22.1-src/src/librustc_plugin)
building of '/nix/store/1q00yw9nphvwvxyk6imxny1rk76rwqy1-rustc-1.22.1.drv' timed out after 3600 seconds
cannot build derivation '/nix/store/i63b6jfgcjz1qkbwsljw4y9vrdsq0qr3-cargo-0.23.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/54gwpxhrss33zvavc8qpv9kvca66ja7j-librsvg-2.42.2.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/4sgr9nibc20a50wqj0m0slhxnb5wsy2d-hook.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/ww752cmphyqnpsqafrhdmq5yzqh3l8v2-icon-naming-utils-0.8.90.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/b1lgn81fw7nxik79zjwshvggx28qgi14-adwaita-icon-theme-3.26.1.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/mka1qxai26bjzqapx40r70c032r96pna-screenkey-0.9.drv': 2 dependencies couldn't be built
error: build of '/nix/store/mka1qxai26bjzqapx40r70c032r96pna-screenkey-0.9.drv' failed
```
</p></details>

vidbina commented:
sure thing... I'm on it now

On Mon, Feb 12, 2018 at 06:07:44PM +0000, Alexey Shmalko wrote:
> @Mic92, updated to include icons.
> 
> @jtojnar, I removed `pythonPackages.screenkey`. (Don't think it deserves an alias as I believe nobody is using it.)
> 
> @vidbina would you like to review/test as well?
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/NixOS/nixpkgs/pull/34167#issuecomment-365009471

vidbina commented:
@rasendubi, on my setup this fails with 

```
DEBUG:Screenkey:Options loaded.
DEBUG:Screenkey:Restarting LabelManager.
DEBUG:Screenkey:Thread started.
DEBUG:Screenkey:Restarting LabelManager.
Exception in thread Thread-1:
Traceback (most recent call last):
  File /nix/store/i3bx1iw2d0i3vh9sa1nf92ynlrw324w8-python-2.7.14/lib/python2.7/threading.py, line 801, in __bootstrap_inner
    self.run()
  File /nix/store/r7dqmjgcp6y72939rwsm0x941k7kd78y-screenkey-0.9/lib/python2.7/site-packages/Screenkey/keylistener.py, line 245, in run
    raise Exception(Cannot initialize input method)
Exception: Cannot initialize input method

zsh: terminated  ./result/bin/screenkey -d
```
It fails on the XopenIM call which I resolved in my old PR ( #34898) by unsetting `XMODIFIERS`. Using `wrapProgram` I kind of know how to do this the "right" way more or less, but with `wrapGAppsHook` I figured appending to `gappsWrapperArgs` would suffice eventhough there aren't any examples of unsetting variables this way in the entire nixpkgs repository so far... or my grep was lousy.

I proposed a [PR](https://github.com/rasendubi/nixpkgs/pull/1) to this PR to keep some work of @rasendubi's shoulders. I understand it's been much 😅, so @jtojnar and @dotlambda please weigh in on this and provide some feedback. I'm not familiar with unsetting variables using the `wrapsGAppsHook` helper, so the proposed may not be idiomatic or "clean" in which case I'd like to know how to clean things up :wink:

jtojnar commented:
Using `gappsWrapperArgs`  is the right way since the setup hook basically just passes it to `wrapProgram` https://github.com/NixOS/nixpkgs/blob/804285f589ded484b530750dd1ec03b9052bcdac/pkgs/build-support/setup-hooks/wrap-gapps-hook.sh#L45. I am not convinced this is the right solution, though. What is the content of the variable? Do other GTK 2 applications work for you?

vidbina commented:
@jtojnar, yeah other GTK 2 apps work for me. In my case `XMODIFIERS` has been set to `@xm=ibus` because I enabled I-Bus as the input method to be able to enter Hiragana, Katakana and Emoji characters.

Enabling an input method [sets `XMODIFIERS`](https://github.com/NixOS/nixpkgs/blob/9b080c35b66abc95c1ad5f0707c9039eb6dcf5d3/nixos/modules/i18n/input-method/ibus.nix#L64) and the modifier var is checked by the i18n xlib [`XSetLocaleModifiers`](https://github.com/mirror/libX11/blob/e835a9dcc3362b5e92893be756dd7ae361e64ced/src/xlibi18n/lcWrap.c#L74-L93) function which is an [important puzzle piece in managing `XOpenIM`'s default behaviour](https://github.com/mirror/libX11/blob/e835a9dcc3362b5e92893be756dd7ae361e64ced/man/XOpenIM.man#L141-L146).

Between input methods such as [fcitx](https://github.com/NixOS/nixpkgs/blob/9b080c35b66abc95c1ad5f0707c9039eb6dcf5d3/nixos/modules/i18n/input-method/fcitx.nix#L39), [ibus](https://github.com/NixOS/nixpkgs/blob/9b080c35b66abc95c1ad5f0707c9039eb6dcf5d3/nixos/modules/i18n/input-method/ibus.nix#L64), [nabi](https://github.com/NixOS/nixpkgs/blob/9b080c35b66abc95c1ad5f0707c9039eb6dcf5d3/nixos/modules/i18n/input-method/nabi.nix#L11) or [uim](https://github.com/NixOS/nixpkgs/blob/9b080c35b66abc95c1ad5f0707c9039eb6dcf5d3/nixos/modules/i18n/input-method/uim.nix#L30) and the method `XopenIM` I'm not exactly sure what the problem is. I see that `XopenIM` calls `_XOpenLC`[:book:](https://github.com/mirror/libX11/blob/0e45f64766c0557c8e99a979c70ca6f55664dae7/specs/i18n/framework/framework.xml#L275-L303) [:octocat:](https://github.com/mirror/libX11/blob/e835a9dcc3362b5e92893be756dd7ae361e64ced/src/xlibi18n/lcWrap.c#L244-L319) which I figure then bumps into an invalid locale loader or a situation where all locale loaders failed to load an locale.

Several threads online proposed the `unset XMODIFIERS` as a quick-fix, there is even another nixpkgs package that does that (some Steam game, I believe) so there is a precedent for this approach.
